# This File is the Exploratory Data Analysis File for the NBA dataset.
Where I will:
1. Check for missing values.
2. Check for duplicates.
3. Check for outliers.
4. handle missing & duplicate values.
5. handle outliers.


# 1. Print information about dataset.

## 1.1. Import all libraries

In [330]:
import sys
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import warnings
from sklearn.preprocessing import LabelEncoder
warnings.filterwarnings('ignore')

## 1.2. Import the dataset

In [331]:
try:
    df = pd.read_csv('nba2k-full.csv')
except FileNotFoundError:
    print("File not found. Please ensure the dataset is in the correct directory.")
    sys.exit(1)
except Exception as e:
    print(f"An error occurred while reading the dataset: {e}")
    sys.exit(1)

In [332]:
df.head(10)

full_name  rating jersey                    team position  \
0           LeBron James      97    #23      Los Angeles Lakers        F   
1          Kawhi Leonard      97     #2    Los Angeles Clippers        F   
2  Giannis Antetokounmpo      96    #34         Milwaukee Bucks      F-G   
3           Kevin Durant      96     #7           Brooklyn Nets        F   
4           James Harden      96    #13         Houston Rockets        G   
5          Stephen Curry      95    #30   Golden State Warriors        G   
6          Anthony Davis      94     #3      Los Angeles Lakers      F-C   
7            Paul George      93    #13    Los Angeles Clippers        F   
8         Damian Lillard      92     #0  Portland Trail Blazers        G   
9            Joel Embiid      91    #21      Philadelphia 76ers        C   

      b_day       height                weight     salary   country  \
0  12/30/84   6-9 / 2.06  250 lbs. / 113.4 kg.  $37436858       USA   
1  06/29/91   6-7 / 2.01  225 lbs. / 102.1 kg.  $32742000       USA   
2  12/06/94  6-11 / 2.11  242 lbs. / 109.8 kg.  $25842697    Greece   
3  09/29/88  6-10 / 2.08  230 lbs. / 104.3 kg.  $37199000       USA   
4  08/26/89   6-5 / 1.96   220 lbs. / 99.8 kg.  $38199000       USA   
5  03/14/88   6-3 / 1.91   185 lbs. / 83.9 kg.  $40231758       USA   
6  03/11/93  6-10 / 2.08  222 lbs. / 100.7 kg.  $27093019       USA   
7  05/02/90   6-8 / 2.03   210 lbs. / 95.3 kg.  $33005556       USA   
8  07/15/90   6-2 / 1.88   195 lbs. / 88.5 kg.  $29802321       USA   
9  03/16/94   7-0 / 2.13  250 lbs. / 113.4 kg.  $27504630  Cameroon   

   draft_year draft_round draft_peak          college  version  
0        2003           1          1              NaN  NBA2k20  
1        2011           1         15  San Diego State  NBA2k20  
2        2013           1         15              NaN  NBA2k20  
3        2007           1          2            Texas  NBA2k20  
4        2009           1          3    Arizona State  NBA2k20  
5        2009           1          7         Davidson  NBA2k20  
6        2012           1          1         Kentucky  NBA2k20  
7        2010           1         10     Fresno State  NBA2k20  
8        2012           1          6      Weber State  NBA2k20  
9        2014           1          3           Kansas  NBA2k20

## 1.3. Print information and shape of dataset

In [333]:
df.shape

(464, 15)

- There are a total of 464 rows and 15 columns in the dataset

In [334]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 464 entries, 0 to 463
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   full_name    464 non-null    object
 1   rating       464 non-null    int64 
 2   jersey       464 non-null    object
 3   team         441 non-null    object
 4   position     464 non-null    object
 5   b_day        464 non-null    object
 6   height       464 non-null    object
 7   weight       464 non-null    object
 8   salary       464 non-null    object
 9   country      464 non-null    object
 10  draft_year   464 non-null    int64 
 11  draft_round  464 non-null    object
 12  draft_peak   464 non-null    object
 13  college      388 non-null    object
 14  version      464 non-null    object
dtypes: int64(2), object(13)
memory usage: 54.5+ KB


- There are 13 numerical columns and 2 categorical columns in the dataset.
- the 'college' & 'team' columns have missing/null values.

## 1.4. Checking for unique values in each column

In [335]:
df['team'].nunique()

30

In [336]:
df['team'].value_counts()

team
Brooklyn Nets             19
Milwaukee Bucks           19
Los Angeles Lakers        18
Dallas Mavericks          18
Los Angeles Clippers      17
Philadelphia 76ers        16
Phoenix Suns              16
Orlando Magic             15
New Orleans Pelicans      15
New York Knicks           15
Indiana Pacers            15
Toronto Raptors           15
Chicago Bulls             15
Denver Nuggets            15
Oklahoma City Thunder     14
Sacramento Kings          14
Detroit Pistons           14
Boston Celtics            14
Minnesota Timberwolves    14
Portland Trail Blazers    14
Houston Rockets           14
Utah Jazz                 14
San Antonio Spurs         14
Washington Wizards        14
Memphis Grizzlies         13
Miami Heat                13
Atlanta Hawks             13
Golden State Warriors     12
Cleveland Cavaliers       12
Charlotte Hornets         10
Name: count, dtype: int64

- Team column has 30 unique teams and can't be dropped or replaced.

In [337]:
df['position'].unique()

array(['F', 'F-G', 'G', 'F-C', 'C', 'G-F', 'C-F'], dtype=object)

In [338]:
df['position'].value_counts()

position
G      184
F      146
C       53
F-C     39
G-F     21
F-G     15
C-F      6
Name: count, dtype: int64

- The position column has 7 unique positions and can't be dropped or replaced.

In [339]:
df['height'].unique()

array(['6-9 / 2.06', '6-7 / 2.01', '6-11 / 2.11', '6-10 / 2.08',
       '6-5 / 1.96', '6-3 / 1.91', '6-8 / 2.03', '6-2 / 1.88',
       '7-0 / 2.13', '6-6 / 1.98', '7-1 / 2.16', '6-1 / 1.85',
       '6-0 / 1.83', '6-4 / 1.93', '7-3 / 2.21', '5-11 / 1.80',
       '5-10 / 1.78', '5-9 / 1.75', '7-4 / 2.24', '7-2 / 2.18'],
      dtype=object)

- Height column displays both in foot and in cm.
- To make the data simpler, we can extract only the cm values from the height column and update the height column.

In [340]:
# Extract only cm values from the height column
df['height_cm'] = (
    df['height']
    .str.split(' / ')    # Split into ['feet-inches', 'meter_value.']
    .str[1]              # Get the second part: 'meter_value.'
    .str.rstrip('.')     # Remove trailing dots: 'meter_value'
    .astype(float)       # Convert to float
    * 100                # Convert meters to centimeters
)
df['height_cm'].astype(int)
df['height_cm'].unique()

array([206., 201., 211., 208., 196., 191., 203., 188., 213., 198., 216.,
       185., 183., 193., 221., 180., 178., 175., 224., 218.])

# 2. Dropping unnecessary columns

- Drop the 'college', 'jersey', and 'full_name' columns

In [341]:
cols_to_drop = ['jersey', 'full_name']
df.drop(columns=cols_to_drop, inplace=True)
df.head(10)

rating                    team position     b_day       height  \
0      97      Los Angeles Lakers        F  12/30/84   6-9 / 2.06   
1      97    Los Angeles Clippers        F  06/29/91   6-7 / 2.01   
2      96         Milwaukee Bucks      F-G  12/06/94  6-11 / 2.11   
3      96           Brooklyn Nets        F  09/29/88  6-10 / 2.08   
4      96         Houston Rockets        G  08/26/89   6-5 / 1.96   
5      95   Golden State Warriors        G  03/14/88   6-3 / 1.91   
6      94      Los Angeles Lakers      F-C  03/11/93  6-10 / 2.08   
7      93    Los Angeles Clippers        F  05/02/90   6-8 / 2.03   
8      92  Portland Trail Blazers        G  07/15/90   6-2 / 1.88   
9      91      Philadelphia 76ers        C  03/16/94   7-0 / 2.13   

                 weight     salary   country  draft_year draft_round  \
0  250 lbs. / 113.4 kg.  $37436858       USA        2003           1   
1  225 lbs. / 102.1 kg.  $32742000       USA        2011           1   
2  242 lbs. / 109.8 kg.  $25842697    Greece        2013           1   
3  230 lbs. / 104.3 kg.  $37199000       USA        2007           1   
4   220 lbs. / 99.8 kg.  $38199000       USA        2009           1   
5   185 lbs. / 83.9 kg.  $40231758       USA        2009           1   
6  222 lbs. / 100.7 kg.  $27093019       USA        2012           1   
7   210 lbs. / 95.3 kg.  $33005556       USA        2010           1   
8   195 lbs. / 88.5 kg.  $29802321       USA        2012           1   
9  250 lbs. / 113.4 kg.  $27504630  Cameroon        2014           1   

  draft_peak          college  version  height_cm  
0          1              NaN  NBA2k20      206.0  
1         15  San Diego State  NBA2k20      201.0  
2         15              NaN  NBA2k20      211.0  
3          2            Texas  NBA2k20      208.0  
4          3    Arizona State  NBA2k20      196.0  
5          7         Davidson  NBA2k20      191.0  
6          1         Kentucky  NBA2k20      208.0  
7         10     Fresno State  NBA2k20      203.0  
8          6      Weber State  NBA2k20      188.0  
9          3           Kansas  NBA2k20      213.0

# 3. Preprocessing

## 3.1. Filling missing values in 'team' column

In [342]:
df['team'].fillna('Free Agent', inplace=True)
df[['team']]

team
0       Los Angeles Lakers
1     Los Angeles Clippers
2          Milwaukee Bucks
3            Brooklyn Nets
4          Houston Rockets
..                     ...
459  Oklahoma City Thunder
460     Washington Wizards
461              Utah Jazz
462        Toronto Raptors
463       Dallas Mavericks

[464 rows x 1 columns]

In [343]:
df['team'].value_counts()

team
Free Agent                23
Brooklyn Nets             19
Milwaukee Bucks           19
Dallas Mavericks          18
Los Angeles Lakers        18
Los Angeles Clippers      17
Philadelphia 76ers        16
Phoenix Suns              16
Orlando Magic             15
New Orleans Pelicans      15
New York Knicks           15
Indiana Pacers            15
Toronto Raptors           15
Chicago Bulls             15
Denver Nuggets            15
Oklahoma City Thunder     14
Sacramento Kings          14
Utah Jazz                 14
Boston Celtics            14
Minnesota Timberwolves    14
Houston Rockets           14
Portland Trail Blazers    14
Detroit Pistons           14
Washington Wizards        14
San Antonio Spurs         14
Memphis Grizzlies         13
Miami Heat                13
Atlanta Hawks             13
Golden State Warriors     12
Cleveland Cavaliers       12
Charlotte Hornets         10
Name: count, dtype: int64

In [344]:
df['team'].isnull().sum()

np.int64(0)

## 3.2. Filling missing values in 'college' column

## 3.3. Splitting & converting birthdate column into birthyear by using the version column

In [345]:
df['b_day'].head(10)

0    12/30/84
1    06/29/91
2    12/06/94
3    09/29/88
4    08/26/89
5    03/14/88
6    03/11/93
7    05/02/90
8    07/15/90
9    03/16/94
Name: b_day, dtype: object

In [346]:
# keep only the year from the 'b_day' column
df['b_year'] = df['b_day'].str.split('/').str[2].astype(int) + 1900
df['b_year'].unique()

array([1984, 1991, 1994, 1988, 1989, 1993, 1990, 1992, 1995, 1996, 1999,
       1985, 1987, 1997, 1998, 1986, 1900, 1980, 1981, 1982])

In [347]:
df.head()

rating                  team position     b_day       height  \
0      97    Los Angeles Lakers        F  12/30/84   6-9 / 2.06   
1      97  Los Angeles Clippers        F  06/29/91   6-7 / 2.01   
2      96       Milwaukee Bucks      F-G  12/06/94  6-11 / 2.11   
3      96         Brooklyn Nets        F  09/29/88  6-10 / 2.08   
4      96       Houston Rockets        G  08/26/89   6-5 / 1.96   

                 weight     salary country  draft_year draft_round draft_peak  \
0  250 lbs. / 113.4 kg.  $37436858     USA        2003           1          1   
1  225 lbs. / 102.1 kg.  $32742000     USA        2011           1         15   
2  242 lbs. / 109.8 kg.  $25842697  Greece        2013           1         15   
3  230 lbs. / 104.3 kg.  $37199000     USA        2007           1          2   
4   220 lbs. / 99.8 kg.  $38199000     USA        2009           1          3   

           college  version  height_cm  b_year  
0              NaN  NBA2k20      206.0    1984  
1  San Diego State  NBA2k20      201.0    1991  
2              NaN  NBA2k20      211.0    1994  
3            Texas  NBA2k20      208.0    1988  
4    Arizona State  NBA2k20      196.0    1989

## 3.4. Splitting game version to extract the year

In [348]:
df['version'].unique()

array(['NBA2k20', 'NBA2k21', 'NBA2k22'], dtype=object)

In [349]:
df['version_split'] = df['version'].str.split('k', expand=True)[1].astype(int)
df['version_split'].unique()

array([20, 21, 22])

- Then convert the 'version_split' to get the release date year

In [350]:
df['version_split'] = df['version_split'] + 2000
df['version_split'].unique()

array([2020, 2021, 2022])

## 3.5. Adding Age column by subtracting birthyear from current game version

In [351]:
df['p_age'] = df['version_split'] - df['b_year']
df['p_age'].head(10)

0    36
1    29
2    26
3    32
4    31
5    32
6    27
7    30
8    30
9    26
Name: p_age, dtype: int64

- This way, we've finally subtracted each player's current age as of the game's version.

## 3.6. Split weight column & extract only the kg values

In [352]:
df['weight_kg'] = (
    df['weight']
    .str.split(' / ')
    .str[1]  # Get the second part (kg value)
    .str.replace(r' kg\.?', '', regex=True)  # Remove 'kg' with optional period
    .astype(float)
)

df['weight_kg'].unique()

array([113.4, 102.1, 109.8, 104.3,  99.8,  83.9, 100.7,  95.3,  88.5,
        81.6, 114.8,  86.2,  93. , 112.5, 108. , 114.3,  95.7,  78. ,
       103. , 111.6,  93.9,  89.4,  96.6,  87.5, 108.9,  77.6,  93.4,
       122.5,  92.5, 102.5,  82.1,  98.4,  79.4,  96.2, 100.2,  91.2,
       106.6,  98. , 101.2, 117.9, 131.5, 120.2, 115.7, 110. , 108.4,
       128.8,  88.9, 104.8,  94.8,  94.3, 118.8,  89.8,  90.7,  78.9,
       110.2, 106.1,  91.6, 109.3, 117.5,  84.4,  97.5,  86.6,  77.1,
        92.1, 101.6,  97.1,  90.3,  87.1, 105.2, 111.1, 103.9,  84.8,
       107.5, 112. ,  99.3,  80.7,  98.9, 107. ,  83.5,  85.7, 104. ,
       112.9,  88. ,  81.2])

## 3.7. Remove '$' sign from salary column

In [353]:
df['salary_usd'] = df['salary'].str.replace(r'\$', '', regex=True).astype(float)
df[['salary_usd']].head(10)

salary_usd
0  37436858.0
1  32742000.0
2  25842697.0
3  37199000.0
4  38199000.0
5  40231758.0
6  27093019.0
7  33005556.0
8  29802321.0
9  27504630.0

## 3.8. Replacing 'undrafted' in 'draft_peak' column with 0

In [354]:
df['draft_peak'].unique()

array(['1', '15', '2', '3', '7', '10', '6', '41', '4', '11', '30', '27',
       '13', '9', '5', '35', '17', '16', '18', '25', '19', '24', '39',
       '45', '31', '36', '32', '48', '12', '38', '47', '8', '33',
       'Undrafted', '40', '14', '22', '42', '46', '20', '29', '26', '51',
       '43', '34', '21', '23', '60', '44', '28', '37', '55', '49', '50',
       '52', '54', '58'], dtype=object)

In [355]:
df['draft_peak'].replace('Undrafted', "0", inplace=True)
df['draft_peak'] = df['draft_peak'].astype(int)
df['draft_peak'].unique()

array([ 1, 15,  2,  3,  7, 10,  6, 41,  4, 11, 30, 27, 13,  9,  5, 35, 17,
       16, 18, 25, 19, 24, 39, 45, 31, 36, 32, 48, 12, 38, 47,  8, 33,  0,
       40, 14, 22, 42, 46, 20, 29, 26, 51, 43, 34, 21, 23, 60, 44, 28, 37,
       55, 49, 50, 52, 54, 58])

## 3.8. Replacing 'undrafted' in 'draft_round' to 0

In [356]:
df['draft_round'].unique()

array(['1', '2', 'Undrafted'], dtype=object)

In [357]:
df['draft_round'] = df['draft_round'].replace('Undrafted', "0")
df['draft_round'] = df['draft_round'].astype(int)
df['draft_round'].unique()

array([1, 2, 0])

## 3.8. Dropping unnecessary columns
- Dropping all the columns I cleaned and we don't need for our model.

In [358]:
cols_to_drop2 = ['height', 'weight', 'b_day', 'version', 'salary', 'college', 'b_year', 'country']
df.drop(columns=cols_to_drop2, inplace=True)
df.head()

rating                  team position  draft_year  draft_round  draft_peak  \
0      97    Los Angeles Lakers        F        2003            1           1   
1      97  Los Angeles Clippers        F        2011            1          15   
2      96       Milwaukee Bucks      F-G        2013            1          15   
3      96         Brooklyn Nets        F        2007            1           2   
4      96       Houston Rockets        G        2009            1           3   

   height_cm  version_split  p_age  weight_kg  salary_usd  
0      206.0           2020     36      113.4  37436858.0  
1      201.0           2020     29      102.1  32742000.0  
2      211.0           2020     26      109.8  25842697.0  
3      208.0           2020     32      104.3  37199000.0  
4      196.0           2020     31       99.8  38199000.0

# 4. Final Preprocessing (Encoding categorical columns)
- Using label encoder to encode the two categorical columns 'team', and 'position'.
- teams are 30 so there will be 30 labels.
- positions are 5 so there will be 5 labels. 

In [359]:
le = LabelEncoder()
df['team_Lencoded'] = le.fit_transform(df['team'])
df[['team_Lencoded']]

team_Lencoded
0               14
1               13
2               17
3                2
4               11
..             ...
459             21
460             30
461             29
462             28
463              6

[464 rows x 1 columns]

In [360]:
df['position_Lencoded'] = le.fit_transform(df['position'])
df[['position_Lencoded']]

position_Lencoded
0                    2
1                    2
2                    4
3                    2
4                    5
..                 ...
459                  5
460                  5
461                  0
462                  5
463                  3

[464 rows x 1 columns]

## 4.1. Dropping the original unencoded columns

In [361]:
df.drop(columns=['team', 'position'], inplace=True)
df.head()

rating  draft_year  draft_round  draft_peak  height_cm  version_split  \
0      97        2003            1           1      206.0           2020   
1      97        2011            1          15      201.0           2020   
2      96        2013            1          15      211.0           2020   
3      96        2007            1           2      208.0           2020   
4      96        2009            1           3      196.0           2020   

   p_age  weight_kg  salary_usd  team_Lencoded  position_Lencoded  
0     36      113.4  37436858.0             14                  2  
1     29      102.1  32742000.0             13                  2  
2     26      109.8  25842697.0             17                  4  
3     32      104.3  37199000.0              2                  2  
4     31       99.8  38199000.0             11                  5

# 5. Print the Dataset so far

In [362]:
df.head()  # Display the first few rows of the dataset

rating  draft_year  draft_round  draft_peak  height_cm  version_split  \
0      97        2003            1           1      206.0           2020   
1      97        2011            1          15      201.0           2020   
2      96        2013            1          15      211.0           2020   
3      96        2007            1           2      208.0           2020   
4      96        2009            1           3      196.0           2020   

   p_age  weight_kg  salary_usd  team_Lencoded  position_Lencoded  
0     36      113.4  37436858.0             14                  2  
1     29      102.1  32742000.0             13                  2  
2     26      109.8  25842697.0             17                  4  
3     32      104.3  37199000.0              2                  2  
4     31       99.8  38199000.0             11                  5

- So far I have cleaned the dataset and removed unnecessary columns and missing values.
- I've also added new columns for 'p_age', 'height_cm', 'weight_kg', and 'version_split' to make the dataset more readable and easier to work with.

In [363]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 464 entries, 0 to 463
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rating             464 non-null    int64  
 1   draft_year         464 non-null    int64  
 2   draft_round        464 non-null    int64  
 3   draft_peak         464 non-null    int64  
 4   height_cm          464 non-null    float64
 5   version_split      464 non-null    int64  
 6   p_age              464 non-null    int64  
 7   weight_kg          464 non-null    float64
 8   salary_usd         464 non-null    float64
 9   team_Lencoded      464 non-null    int64  
 10  position_Lencoded  464 non-null    int64  
dtypes: float64(3), int64(8)
memory usage: 40.0 KB


## 5.1. Save the cleaned dataset to a new CSV file (version 1)

In [364]:
df.to_csv('nba2k_cleaned_v1.csv', index=False)

# 6. Conclusion

- Cleaned the dataset completely from missing values and made sure no duplicates are present.
- Added new columns for 'p_age', 'height_cm', 'weight_kg', and 'version_split' to make the dataset more readable and easier to work with.
- Dropped unnecessary columns that won't affect the player's salary (irrelevant).
- Used Label Encoding to Encode the 'team' and 'position' columns which where the only two remaining categorical columns.
- Saved the cleaned dataset to a new CSV file (version 1 for testing).